### ISSUE TITLE
fs events ops

### ISSUE THREAD

**Issue by kitsonk** [denoland/deno#1826]
There are several use cases where the ability to watch a file(s) for changes is an important API.  For example, invalidating cache headers for serving static file based content, or processing logs or files, etc.

Many OSes provide a mechanism for this without having to "poll" the stats for the file.  Node.js offers [`fs.watch()`](https://nodejs.org/docs/latest/api/fs.html#fs_fs_watch_filename_options_listener).

Instead of an event listener though, it feels some sort of async generator would be more idiomatic to the rest of Deno.

**Comment by ry** [comment]
Yes would have in core. This looks pretty nice: https://github.com/passcod/notify

**Comment by ry** [comment]
Here's a potential API, but I leave the exact spelling of these methods to the implementor
```
  const watcher = await Deno.watch("/home/test/notify", { recursive: true });
  while (true) {
    const event = await watcher.getEvent();
    console.log("event", event)
  }
```
Here's a brief sketch of how I would do this

1. Add new ops to [cli/msg.fbs ](https://github.com/denoland/deno/blob/5ae78eb1de378f04d0b9d54842bcb898053467d6/cli/msg.fbs)
1A. One to create a new watcher, maybe FSEventsOpen. This would return a new resource id (rid).
1B. One to poll for a new event, FSEventsPoll. It would take an rid and return a new event.
1C. Closing the rid does not need a new op. The existing Close op can be used.
2. Add new handlers for the two new ops in [ops.rs](https://github.com/denoland/deno/blob/5ae78eb1de378f04d0b9d54842bcb898053467d6/cli/ops.rs#L151-L199)
2A. Add a new Resource type in [resources.rs](https://github.com/denoland/deno/blob/5ae78eb1de378f04d0b9d54842bcb898053467d6/cli/resources.rs#L79-L102), you can probably follow the example of `tokio_process::Child`
2B. Add a new typescript frontend to the op, by adding `js/fs_events.ts`, follow the example in `js/process.ts`.
3. Write some tests.

**Comment by balupton** [comment]
Having written github.com/bevry/watchr I would encourage no built in watching apis, as watching is such a difficult problem. I think in nearly all cases people would be best served by facebook's watchman instead. Which does not use the node apis at all.

**Comment by ry** [comment]
watching != fs notifications

**Comment by sholladay** [comment]
Personally, I've frequently been annoyed at how file watching was a mess in Node, a situation that is still unresolved today. Yes, it's hard to implement correctly, but that just means we need more eyes on it. Also, a lot of projects need it as a very low-level building block. For context, [chokidar](https://github.com/paulmillr/chokidar) has almost 11 million weekly downloads on npm, about as much as React and Express combined (i.e. it is among the most downloaded packages).

I would love to see this in core and exposed as an async iterable.

```js
const watcher = Deno.watch('/home/test/notify', { recursive: true });
for await (const event of watcher) {
    // ...
}
```

The snippet above is using [for-await-of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of). Node recently started to use this for streams and I've fallen in love with it. It's standard syntax, which aligns nicely with Deno's philosophy.

**Comment by jcao219** [comment]
I can try to see if I can make a prototype of @sholladay 's API into reality.

**Comment by Fenzland** [comment]
I'm waiting for this feature too. Using `for-await-of` is pretty nice. But in some using cases, we need to listen to events with different types. Nesting several `if` or `switch`es in the `for` is not so good. I have a plan to combine both event listener and for-await-of together:

```javascript
const watcher = Deno.watch('/path', { recursive: true });
watcher instanceof Watcher;
watcher.addEventListener('add', onAdd);
watcher.addEventListener('modify', ()=> {});
watcher.addEventListener('chmod', ()=> {});
watcher.removeEventListener('add', onAdd)

for await ( let { path } of watcher.onAdd() ) {
  //...
}
```

**Comment by MarkTiedemann** [comment]
> What's more, A close() API is necessary.

I think if we use async generators, then we don't need a `close` API, but can use the built-in `return` API (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator/return).

```typescript
let watcher = Deno.watch(/*...*/);
// ...

// some time later
await watcher.return(); // wait for watcher to close
```

**Comment by Fenzland** [comment]
> I think if we use async generators, then we don't need a `close` API, but can use the built-in `return` API 

That's right. But now, I change my mind and not support for-await-of any more.

**Comment by ry** [comment]
Added to 1.0 blockers

**Comment by bartlomieju** [comment]
@jcao219 I remember you were working on an implementation, any update?

**Comment by jcao219** [comment]
@bartlomieju Hi all, sorry for the long delay.

I have had a working implementation, but it's been a while since I've synced it with master branch, so I'll need to update it for the new JSON message passing changes. Let me do that ASAP and get right back to y'all.

**Comment by Dan503** [comment]
It would be nice if the watcher could tell the difference between adding/removing a file and simply renaming it.

Chockidar can't tell the difference between renaming and adding/removing so when you rename something it will fire both an "add" event and a "remove" event. This can lead to tasks being run twice when you rename a file.

**Comment by bartlomieju** [comment]
Ref #3452

**Comment by bartlomieju** [comment]
This feature is almost complete (#3452), only common interface must be established. I'm marking this issue as "help wanted" - if anyone wants to pick up my PR feel free.

**Comment by ry** [comment]
Fixed in #3452
